SimG4Core/Application/src/TrackingAction.cc

```diff
@@ -38,21 +40,71 @@ void TrackingAction::PreUserTrackingAction(const G4Track* aTrack) {
   if (nullptr != steppingVerbose_) {
     steppingVerbose_->TrackStarted(aTrack, false);
   }
+
+  if (doFineCalo_) {
+#ifdef EDM_ML_DEBUG
+    edm::LogVerbatim("DoFineCalo") << "PreUserTrackingAction: Start processing track " << aTrack->GetTrackID()
+                                   << " pdgid=" << aTrack->GetDefinition()->GetPDGEncoding()
+                                   << " ekin[GeV]=" << aTrack->GetKineticEnergy() / CLHEP::GeV << " vertex[cm]=("
+                                   << aTrack->GetVertexPosition().x() / CLHEP::cm << ","
+                                   << aTrack->GetVertexPosition().y() / CLHEP::cm << ","
+                                   << aTrack->GetVertexPosition().z() / CLHEP::cm << ")"
+                                   << " parentid=" << aTrack->GetParentID();
+#endif
+    // It is impossible to tell whether daughter tracks if this track may need to be saved at
+    // this point; Therefore, *every* track is put in history, so that it can potentially be saved
+    // later.
+    trkInfo->putInHistory();
+    // Always save primaries
+    // Decays from primaries are marked as primaries (see NewTrackAction), but are not saved by
+    // default. The primary is the earliest ancestor, and it must be saved.
+    if (trkInfo->isPrimary())
+      currentTrack_->save();
+  }
 }
 
 void TrackingAction::PostUserTrackingAction(const G4Track* aTrack) {
   if (eventAction_->trackContainer() != nullptr) {
-    if (extractor_(aTrack).storeTrack()) {
```

Comment:
there may be reasons to store track without any further check?

Reply:
This line is just moved to line 105 and I added an extra check check to see if the TrackWithHistory flag was set manually: `if (extractor_(aTrack).storeTrack() || currentTrack_->saved()) {`.